jit/jit.cpp

```diff
@@ -0,0 +1,467 @@
+    }
+    auto* jit_symbol_address = reinterpret_cast<void*>(jit_symbol.getAddress());
+    compiled_module.function_name_to_symbol.emplace(std::move(function_name), jit_symbol_address);
+  }
+  compiled_module.size = module_memory_manager->allocatedMemorySize();
+  compiled_module.identifier = current_module_key;
+  module_identifier_to_memory_manager.emplace(current_module_key, std::move(module_memory_manager));
+  compiled_code_size.fetch_add(compiled_module.size, std::memory_order_relaxed);
+  return compiled_module;
+}
+
+std::unique_ptr<llvm::Module> JIT::createModuleForCompilation()
+{
+  auto module = std::make_unique<llvm::Module>("jit_module_" + std::to_string(current_module_key), context);
+  module->setDataLayout(data_layout);
+  module->setTargetTriple(target_machine->getTargetTriple().str());
+  return module;
+}
+std::string JIT::getMangledName(const std::string& name_to_mangle) const
+{
+  std::string mangled_name;
+  llvm::raw_string_ostream ostream(mangled_name);
+  llvm::Mangler::getNameWithPrefix(ostream, name_to_mangle, data_layout);
+  ostream.flush();
+
+  return mangled_name;
+}
+std::unique_ptr<llvm::TargetMachine> JIT::getTargetMachine()
+{
+  static std::once_flag llvm_target_machine_initialized;
+  std::call_once(llvm_target_machine_initialized,
+                 []()
+                 {
+                   llvm::InitializeNativeTarget();
+                   llvm::InitializeNativeTargetAsmPrinter();
+                   llvm::sys::DynamicLibrary::LoadLibraryPermanently(nullptr);
+                 });
+  std::string error;
+  auto cpu = llvm::sys::getHostCPUName();
+  auto triple = llvm::sys::getProcessTriple();
+  const auto* target = llvm::TargetRegistry::lookupTarget(triple, error);
+  if (!target)
+  {
+    throw std::runtime_error("Failed to lookup target: " + error);
+  }
+  llvm::SubtargetFeatures features;
+  llvm::StringMap<bool> feature_map;
+  if (llvm::sys::getHostCPUFeatures(feature_map))
+    for (auto& f : feature_map)
+      features.AddFeature(f.first(), f.second);
+
+  llvm::TargetOptions options;
+
+  bool jit = true;
+  auto* target_machine = target->createTargetMachine(triple, cpu, features.getString(), options, llvm::None,
+                                                     llvm::None, llvm::CodeGenOpt::Aggressive, jit);
+
+  if (!target_machine)
+    throw std::runtime_error("Failed to create target machine");
+
+  return std::unique_ptr<llvm::TargetMachine>(target_machine);
+}
+
+void JIT::deleteCompiledModule(JIT::CompiledModule& module)
+{
+  std::lock_guard<std::mutex> lock(jit_lock);
+  auto module_memory_manager_it = module_identifier_to_memory_manager.find(module.identifier);
+  if (module_memory_manager_it == module_identifier_to_memory_manager.end())
+  {
+    throw std::runtime_error("There is no compiled module with identifier");
+  }
+  module_identifier_to_memory_manager.erase(module_memory_manager_it);
+  compiled_code_size.fetch_sub(module.size, std::memory_order_relaxed);
```

Comment:
Which code or data structure owns the memory that `module` referes to here? Could the memory actually be freed in the `erase` call? If so, then these statements should be swapped with each other to make the code safe.